src/Component/Symfony/Request/State/Provider.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Resource\Symfony\Request\State;
+
+use Psr\Container\ContainerInterface;
+use Sylius\Component\Resource\Context\Context;
+use Sylius\Component\Resource\Context\Option\RequestOption;
+use Sylius\Component\Resource\Metadata\CollectionOperationInterface;
+use Sylius\Component\Resource\Metadata\Operation;
+use Sylius\Component\Resource\Reflection\CallableReflection;
+use Sylius\Component\Resource\State\ProviderInterface;
+use Sylius\Component\Resource\Symfony\Request\RepositoryArgumentResolver;
+
+final class Provider implements ProviderInterface
+{
+    public function __construct(
+        private ContainerInterface $locator,
+        private RepositoryArgumentResolver $argumentResolver,
+    ) {
+    }
+
+    public function provide(Operation $operation, Context $context): object|iterable|null
+    {
+        $request = $context->get(RequestOption::class)?->request();
+        $repository = $operation->getRepository();
+
+        if (
+            null === $request ||
+            null === $repository
+        ) {
+            return null;
+        }
+
+        if (\is_string($repository)) {
+            $defaultMethod = $operation instanceof CollectionOperationInterface ? 'createPaginator' : 'findOneBy';
```

Comment:
I would recommend extracting this behaviour outside of this class. Otherwise, change of default method for whatever reason will be harder